src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,399 @@
+                if (hasInvalidOperation)
+                {
+                    return;
+                }
+
+                // Report diagnostics for unused candidate members.
+                var first = true;
+                PooledHashSet<ISymbol> symbolsReferencedInDocComments = null;
+                try
+                {
+                    var namedType = (INamedTypeSymbol)symbolEndContext.Symbol;
+                    foreach (var member in namedType.GetMembers())
+                    {
+                        // Check if the underlying member is neither read nor a readable reference to the member is taken.
+                        // If so, we flag the member as either unused (never written) or unread (written but not read).
+                        if (TryRemove(member, out var valueUsageInfo) &&
+                            !valueUsageInfo.ContainsReadOrReadableRef())
+                        {
+                            Debug.Assert(IsCandidateSymbol(member));
+                            Debug.Assert(!member.IsImplicitlyDeclared);
+
+                            if (first)
+                            {
+                                // Bail out if there are syntax errors in any of the declarations of the containing type.
+                                // Note that we check this only for the first time that we report an unused or unread member for the containing type.
+                                if (HasSyntaxErrors(namedType, symbolEndContext.CancellationToken))
+                                {
+                                    return;
+                                }
+
+                                // Compute the set of candidate symbols referenced in all the documentation comments within the named type declarations.
+                                // This set is computed once and used for all the iterations of the loop.
+                                symbolsReferencedInDocComments = GetCandidateSymbolsReferencedInDocComments(namedType, symbolEndContext.Compilation, symbolEndContext.CancellationToken);
+                                first = false;
+                            }
+
+                            // Report IDE0051 or IDE0052 based on whether the underlying member has any Write/WritableRef/NonReadWriteRef references or not.
+                            var rule = !valueUsageInfo.ContainsWriteOrWritableRef() && !valueUsageInfo.ContainsNonReadWriteRef() && !symbolsReferencedInDocComments.Contains(member)
+                                ? _removeUnusedMembersRule
+                                : _removeUnreadMembersRule;
+                            var effectiveSeverity = rule.GetEffectiveSeverity(symbolEndContext.Compilation.Options);
+
+                            var diagnostic = DiagnosticHelper.Create(
+                                rule,
+                                member.Locations[0],
```

Comment:
won't this squiggle the entire member?  or is .Locations just hte name?  if so, this is fine, but then 'additionalLocations' is not.

Reply:
It squiggles/fades just the symbol name. Additional locations should almost always be null here, except for partial methods, so I am just going to ignore it for now and pass null.

---
In reply to: [213883172](https://github.com/dotnet/roslyn/pull/29511#discussion_r213883172) [](ancestors = 213883172)